File: LLD/InterviewExamples/Comment-Service/src/main/java/com/example/commentservice/services/CommentService.java
package com.example.commentservice.services;

import com.example.commentservice.dto.CommentDTO;
import com.example.commentservice.enums.Comment_Status;
import com.example.commentservice.models.Comment;
import com.example.commentservice.models.CommentLike;
import com.example.commentservice.models.User;
import com.example.commentservice.repositories.CommentLikeRepository;
import com.example.commentservice.repositories.CommentRepository;
import com.example.commentservice.repositories.UserRepository;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;

import org.springframework.data.domain.Pageable;
import java.util.List;

@Service
public class CommentService {

    private CommentRepository commentRepository;
    private UserRepository userRepository;
    private CommentLikeRepository commentLikeRepository;

    public CommentService(CommentRepository commentRepository, UserRepository userRepository,  CommentLikeRepository commentLikeRepository) {
        this.commentRepository = commentRepository;
        this.commentLikeRepository = commentLikeRepository;
        this.userRepository = userRepository;
    }

    public List<Comment> findAllComments(int offset) {
        Pageable pageable = PageRequest.of(offset,10);
        return this.commentRepository.findAll(pageable).getContent();
    }

    public Comment createComment(CommentDTO commentDTO) {
        Comment comment = new Comment(userRepository.findById(commentDTO.getUser_id()).orElseThrow(() -> new IllegalStateException("User is not present")), commentDTO.getContent());
        return this.commentRepository.save(comment);
    }

    public void deleteComment(Long id) {
        Comment comment = commentRepository.findById(id).orElseThrow(() -> new IllegalStateException("Comment doesn't exist"));

        if(comment.getStatus() == Comment_Status.DELETED) {
            throw new IllegalStateException("Comment already deleted");
        } else {
            comment.setStatus(Comment_Status.DELETED);
            commentRepository.save(comment);
        }
    }

    public Comment findCommentById(Long id) {
        return this.commentRepository.findById(id).orElseThrow(() -> new IllegalStateException("Comment not found"));
    }

    public void likeComment(Long comment_id, Long user_id) {
        User user = userRepository.findById(user_id)
                .orElseThrow(() -> new IllegalStateException("User not found"));
        Comment comment = commentRepository.findById(comment_id)
                .orElseThrow(() -> new IllegalStateException("Comment not found"));

        if(commentLikeRepository.existsByCommentAndUser(comment, user)) {
            throw new IllegalStateException("User has already liked the comment");
        }

        commentLikeRepository.save(new CommentLike(user, comment));
        comment.setLike_count(comment.getLike_count() + 1);
        commentRepository.save(comment);
    }

    public void unlikeComment(Long comment_id, Long user_id) {
        User user = userRepository.findById(user_id)
                .orElseThrow(() -> new IllegalStateException("User not found"));
        Comment comment = commentRepository.findById(comment_id)
                .orElseThrow(() -> new IllegalStateException("Comment not found"));

        CommentLike like = commentLikeRepository.findByCommentAndUser(comment, user)
                .orElseThrow(() -> new IllegalStateException("User has not liked the comment"));

        commentLikeRepository.deleteById(like.getId());
        comment.setLike_count(comment.getLike_count() - 1);
        commentRepository.save(comment);
    }

    public int getCommentLikeCount(Long comment_id) {
        Comment comment = commentRepository.findById(comment_id).orElseThrow(() -> new IllegalStateException("Comment is not present"));
        return comment.getLike_count();
    }

}

File: LLD/InterviewExamples/Comment-Service-Node/services/CommentService.js
const conn = require('../config/db')
const queries = require('../queries/Queries')
module.exports = {
    // List all comments with offset
    getComments: async (offset) => {
        const [rows] = await conn.query(queries.getComments(), [Number(offset)]);
        return rows;
    },

    // Create a comment (matching your DTO logic)
    createComment: async (commentDTO) => {
        // Check if user exists (Optional but matches your Java orElseThrow)
        const [user] = await conn.query(queries.getUserById(), [commentDTO.user_id]);
        if (user.length === 0) throw new Error("User is not present");

        const [result] = await conn.query(queries.createComment(), [commentDTO.user_id, commentDTO.content]);
        return result;
    },

    // Get single comment
    getCommentById: getCommentById,

    // Soft delete logic
    deleteComment: async (comment_id) => {
        // MUST AWAIT here to get the actual array
        const rows = await getCommentById(comment_id);
        
        if (rows.length === 0) {
            throw new Error("Comment doesn't exist");
        }
        if (rows[0].status === 'DELETED') {
            throw new Error("Comment already deleted");
        }

        return await conn.query(queries.deleteComment(), [comment_id]);
    },

    // Like logic with check
    likeComment: async (user_id, comment_id) => {
        // 1. Verify user and comment exist
        const user = await conn.query(queries.getUserById(), [user_id]);
        const commentRows = await getCommentById(comment_id);

        if (user.length === 0) throw new Error("User not found");
        if (commentRows.length === 0) throw new Error("Comment not found");

        // 2. Check if already liked (existsByCommentAndUser)
        const [existingLike] = await conn.query(queries.getLikeCommentByUserIdAndCommentId(), [user_id, comment_id]);
        if (existingLike.length > 0) {
            throw new Error("User has already liked the comment");
        }

        // 3. Save like and increment count (Transactions recommended here)
        await conn.query(queries.likeComment(), [user_id, comment_id]);
        const likeCount = await getCommentLikeCount(comment_id)
        return await conn.query(queries.setCommentLikeCount(), [likeCount+1,comment_id]);
    },

    // Unlike logic
    unlikeComment: async (user_id, comment_id) => {
        const [existingLike] = await conn.query(queries.getLikeCommentByUserIdAndCommentId(), [user_id, comment_id]);
        if (existingLike.length === 0) {
            throw new Error("User has not liked the comment");
        }

        await conn.query(queries.unlikeComment(), [user_id, comment_id]);
        const likeCount = await getCommentLikeCount(comment_id)
        return await conn.query(queries.setCommentLikeCount(), [likeCount - 1 ,comment_id]);
    },

    getCommentLikeCount: getCommentLikeCount
};

// Helper functions (Hoisted)
async function getCommentById(comment_id) {
    const [rows] = await conn.query(queries.getCommentById(), [comment_id]);
    return rows; 
}

async function getCommentLikeCount(comment_id) {
    const [rows] = await conn.query(queries.getCommentLikeCount(), [comment_id]);
    if (rows.length === 0) throw new Error("Comment is not present");
    return rows[0].like_count;
}
